Add tests for Dolphin, crab and butterfly

JorVLP opened the pull request:
- Fixes #86 crab no longer can copy from itself, and always has minimum 1 health
- Add tests for dolphin successfully killing caterpillar before evolve, and after evolve but before butterfly gains stats
- Fix abilities triggering after death when they shouldn't
- Pass newly added tests

Let me know if you'd like anything changed, my first time looking at the codebase so I'm not too familiar with all the inner workings.

A few tests outside of test_effect_queue.py are failing, however I've left them as is as they were also failing previously.

manny405 commented on sapai/battle.py:
```diff
@@ -611,14 +611,19 @@ def loop_effect_queue(effect_queue, summon_queue, run_status_summon, init_knocko
-                append_phase_list(
-                    phase_list, p, team_idx, pet_idx, activated, targets, possible
-                )
+                    targets = []
```
Set `activated=False` and `possible=[]` here as well

manny405 reviewed (changes requested):
Thanks for the contributions!

Linting test failed. Linting is done with `black==22.6.0`. In the repo directory, run `black */*.py` from command line and then push the changes. 

You're correct. For now, all other tests that are not in `test_effect_queue` are not of concern. These will be fixed once all targets in `test_effect_queue` are completed to prepare merging this branch with `main`.

Would you mind fixing the rooster example in #86 as well and adding a test? If not, it can be fixed later this PR can certainly be merged.

JorVLP commented:
Added tests and fixed functionality for the rooster.

I noticed there isn't a requirements file in the repository, do you mind if I commit my poetry lock files?

manny405 reviewed (approved):
lgtm
